File: source/Hollow.cpp
/*
* Hollow - A lightweight C++ core library
 * Copyright (c) 2026 VerdantMoss
 *
 * Licensed under the MIT License. See LICENSE file for details.
 *
 * Code Style: Import style.xml for JetBrains IDE formatting
 */
#include "Hollow.hpp"
#include <utility>
#include <locale>
#include <ctime>
#ifdef _WIN32
#include <windows.h>
#else
#include <sys/utsname.h>
#endif
namespace hollow {
	const Version hollowVersion(0, 1, 3, 0, ReleaseType::STABLE);
	const std::string hollowName = "Hollow";
	const Color hollowColor(143, 0, 255);
	static const Hollow internalHollow(hollowVersion, hollowName, hollowColor);
	const Hollow* Hollow::hollow = &internalHollow;
	static bool systemInfoLogged = []() {
		Hollow::logSystemInfo();
		return true;
	}();

	Hollow::Hollow(const Version& version, std::string projectName, const Color& projectColor) : version(version),
		projectName(std::move(projectName)),
		projectColor(projectColor),
		aurora(*this) {
	}

	const Version& Hollow::getVersion() const {
		return version;
	}

	const std::string& Hollow::getProjectName() const {
		return projectName;
	}

	const Color& Hollow::getProjectColor() const {
		return projectColor;
	}

	Aurora& Hollow::getAurora() const {
		return aurora;
	}

	std::string Hollow::getProjectInfo() const {
		return projectName + " (" + version.toString() + ")";
	}

	void Hollow::logSystemInfo() {
		const Aurora& aurora = hollow->getAurora();
		aurora.system("=== System Information ===");
		#ifdef _WIN32
		OSVERSIONINFOEX osvi; ZeroMemory(&osvi, sizeof(OSVERSIONINFOEX)); osvi.dwOSVersionInfoSize = sizeof(
			OSVERSIONINFOEX); if (GetVersionEx((OSVERSIONINFO*)&osvi)) {
			aurora.system(
				"Operating System: Windows " + std::to_string(osvi.dwMajorVersion) + "." + std::to_string(
					osvi.dwMinorVersion));
		}
		#else
		utsname unameData{};
		if (uname(&unameData) == 0) {
			aurora.system("Operating System: " + std::string(unameData.sysname) + " " + std::string(unameData.release));
		}
		#endif
		const char* user = std::getenv("USER");
		if (const char* home = std::getenv("HOME"); user && home) {
			aurora.system("User: " + std::string(user) + " | Home: " + std::string(home));
		}
		try {
			const std::locale loc("");
			std::string localeName = loc.name();
			if (localeName == "C" || localeName.empty()) {
				const char* lang = std::getenv("LANG");
				if (const char* lc_all = std::getenv("LC_ALL"))
					localeName = lc_all;
				else if (lang)
					localeName = lang;
				else
					localeName = "Unknown";
			}
			if (localeName.find(';') != std::string::npos) {
				if (size_t start = localeName.find('='); start != std::string::npos) {
					start++;
					const size_t end = localeName.find(';', start);
					localeName = localeName.substr(start, end - start);
				}
			}
			aurora.system("Locale: " + localeName);
		}
		catch (...) {
			const char* lang = std::getenv("LANG");
			const char* lc_all = std::getenv("LC_ALL");
			std::string localeName = "Unknown";
			if (lc_all)
				localeName = lc_all;
			else if (lang)
				localeName = lang;
			if (localeName.find('.') != std::string::npos) {
				localeName = localeName.substr(0, localeName.find('.'));
			}
			aurora.system("Locale: " + localeName);
		}
		const std::time_t now = std::time(nullptr);
		if (const std::tm* localTime = std::localtime(&now)) {
			char tzBuf[32];
			std::strftime(tzBuf, sizeof(tzBuf), "%Z", localTime);
			aurora.system("Time Zone: " + std::string(tzBuf));
		}
		aurora.system("===========================");
	}

	Hollow& Hollow::getInternalHollow() {
		return *const_cast<Hollow*>(hollow);
	}
}

File: src/main/java/verdant_moss/hollow/Hollow.java
package verdant_moss.hollow;

import verdant_moss.hollow.aurora.Aurora;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Locale;

/**
 * The central class of the Hollow library.
 *
 * <p>
 * Hollow serves mainly as a container for the {@link Aurora} logger and the versioning
 * information.
 * </p>
 */
public class Hollow {
	public static final String HOLLOW_NAME = "Hollow";
	public static final Version HOLLOW_VERSION = new Version(0, 1, 0, 0, ReleaseType.RELEASE);
	public static final Color HOLLOW_COLOR = new Color(101, 0, 179);
	public static final Hollow HOLLOW = new Hollow(HOLLOW_NAME, HOLLOW_VERSION, HOLLOW_COLOR);
	public static final Aurora HOLLOW_AURORA = HOLLOW.get_aurora();
	private final String name;
	private final Version version;
	private final Color color;
	private final Aurora aurora;
	
	public Hollow(String name, Version version, Color color) {
		this.name = name;
		this.version = version;
		this.color = color;
		this.aurora = new Aurora(name, version, color);
	}
	
	/**
	 * Logs system information at startup.
	 */
	private static void logStartupInfo() {
		DateTimeFormatter formatter = DateTimeFormatter.ofPattern("HH.mm.ss-dd.MM.yyyy");
		String timestamp = LocalDateTime.now().format(formatter);
		HOLLOW_AURORA.system("=== Hollow System Information ===");
		HOLLOW_AURORA.system("Time: " + timestamp);
		HOLLOW_AURORA.system("OS: " + System.getProperty("os.name") + " " + System.getProperty("os.version"));
		HOLLOW_AURORA.system("Locale: " + Locale.getDefault());
		HOLLOW_AURORA.system("===========================");
	}
	
	public Aurora get_aurora() {
		return aurora;
	}
	
	public String get_name() {
		return name;
	}
	
	public Version get_version() {
		return version;
	}
	
	public Color get_color() {
		return color;
	}
	
	static {
		logStartupInfo();
	}
}